### PR TITLE
Polymorphismize ships

### LINKED ISSUE #42: Polymorphize ships
I learned that it can be done and it will help our sprint 4. I am completing this ASAP tonight.

### PR DESCRIPTION
**Pull request by astra137**
This should simplify implementing submarine and movement user stories.

Resolves #42

### CODE REVIEW

File: src/main/java/cs361/battleships/models/ships/Ship.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import cs361.battleships.models.Square;
+
+import java.util.List;
+
+@JsonTypeInfo(
+        use = JsonTypeInfo.Id.NAME,
+        property = "type"
+)
+
+@JsonSubTypes({
```
Comment (astra137) [inline src/main/java/cs361/battleships/models/ships/Ship.java:17]: A new subtype will be necessary to support Submarine.

**Comment by astra137** [comment]
The test and JS changes are just to accommodate the new ship classes.

File: src/main/java/cs361/battleships/models/Board.java [hunk 2]
```diff
@@ -1,122 +1,127 @@
...
+    private int hitsAllowed (Square sq) {
+        for (var ship : ships) {
+            if (sq.equals(ship.getCaptainsQuarters())) {
+                if (ship.isCaptainsReinforced()) {
+                    return 2;
+                }
+            }
+        }
+        return 1; // Miss and regular squares
+    }
```
Comment (mi11y) [inline src/main/java/cs361/battleships/models/Board.java:125]: I assume all these lines were touched from just a copy paste?

File: src/main/java/cs361/battleships/models/ships/Ship.java (hunk 1 above)
Comment (mi11y) [reply src/main/java/cs361/battleships/models/ships/Ship.java:17]: I appreciate the abstraction of Ship <3

File: src/test/java/cs361/battleships/models/GameTest.java [hunk 3]
```diff
@@ -1,26 +1,44 @@
 package cs361.battleships.models;
 
+import cs361.battleships.models.ships.Destroyer;
+import cs361.battleships.models.ships.Minesweeper;
+import cs361.battleships.models.ships.Ship;
 import org.junit.Test;
 
+import static junit.framework.TestCase.assertEquals;
+import static junit.framework.TestCase.assertTrue;
```
Comment (mi11y) [inline src/test/java/cs361/battleships/models/GameTest.java:9]: Normal junit doesnt come with assertTrue and Equals?

**Review by mi11y — APPROVED** [review]
For the most part the changes make sense, especially with abstracting ship. What is interesting is that now ships know about the squares they are on by calculating them... which shouldnt that have been the case from the beginning? That feels more OO design that what we were doing before.

Something I do have a question about is in the `placeShip()` method in Game.java. So now each time the method is called, it gets passed the newly placed ships from both the player and enemy?

**Comment by mi11y** [comment]
Review done, comments submitted and tests passed <3

**Pull request merged**